File: ios-app/AnagramArena/Sources/Core/Engine/ValidationAndScoring.swift
import Foundation

protocol DictionaryProviding {
    func contains(_ normalizedWord: String) -> Bool
}

struct WordScorer {
    func scoreLettersWord(length: Int) -> Int {
        guard length > 0 else { return 0 }
        return length == 9 ? 12 : length
    }
}

struct WordValidator {
    private let dictionary: DictionaryProviding
    private let checker: WordConstructabilityChecker
    private let scorer: WordScorer

    init(
        dictionary: DictionaryProviding,
        checker: WordConstructabilityChecker = WordConstructabilityChecker(),
        scorer: WordScorer = WordScorer()
    ) {
        self.dictionary = dictionary
        self.checker = checker
        self.scorer = scorer
    }

    func validate(word rawWord: String, against letters: [Character]) -> WordValidationResult {
        let normalized = WordNormalizer.normalize(rawWord)

        guard !normalized.isEmpty else {
            return WordValidationResult(normalizedWord: normalized, isValid: false, failure: .empty, score: 0)
        }

        guard WordNormalizer.isAlphabetical(normalized) else {
            return WordValidationResult(normalizedWord: normalized, isValid: false, failure: .nonAlphabetical, score: 0)
        }

        guard dictionary.contains(normalized) else {
            return WordValidationResult(normalizedWord: normalized, isValid: false, failure: .notInDictionary, score: 0)
        }

        guard checker.canConstruct(word: normalized, from: letters) else {
            return WordValidationResult(normalizedWord: normalized, isValid: false, failure: .notConstructable, score: 0)
        }

        let score = scorer.scoreLettersWord(length: normalized.count)
        return WordValidationResult(normalizedWord: normalized, isValid: true, failure: nil, score: score)
    }
}

struct ConundrumValidator {
    func isCorrect(guess: String, answer: String) -> Bool {
        WordNormalizer.normalize(guess) == WordNormalizer.normalize(answer)
    }
}

File: android-app/core/src/main/kotlin/com/bhan796/anagram/core/validation/ValidationAndScoring.kt
package com.bhan796.anagram.core.validation

import com.bhan796.anagram.core.model.WordValidationFailure
import com.bhan796.anagram.core.model.WordValidationResult

interface DictionaryProvider {
    fun contains(normalizedWord: String): Boolean
}

class InMemoryDictionaryProvider(private val words: Set<String>) : DictionaryProvider {
    override fun contains(normalizedWord: String): Boolean = words.contains(normalizedWord)
}

class WordScorer {
    fun scoreLettersWord(length: Int): Int {
        if (length <= 0) return 0
        return if (length == 9) 12 else length
    }
}

class WordValidator(
    private val dictionary: DictionaryProvider,
    private val checker: WordConstructabilityChecker = WordConstructabilityChecker(),
    private val scorer: WordScorer = WordScorer()
) {
    fun validate(word: String, letters: List<Char>): WordValidationResult {
        val normalized = WordNormalizer.normalize(word)

        if (normalized.isEmpty()) {
            return WordValidationResult(normalizedWord = normalized, isValid = false, failure = WordValidationFailure.EMPTY, score = 0)
        }

        if (!WordNormalizer.isAlphabetical(normalized)) {
            return WordValidationResult(normalizedWord = normalized, isValid = false, failure = WordValidationFailure.NON_ALPHABETICAL, score = 0)
        }

        if (!dictionary.contains(normalized)) {
            return WordValidationResult(normalizedWord = normalized, isValid = false, failure = WordValidationFailure.NOT_IN_DICTIONARY, score = 0)
        }

        if (!checker.canConstruct(normalized, letters)) {
            return WordValidationResult(normalizedWord = normalized, isValid = false, failure = WordValidationFailure.NOT_CONSTRUCTABLE, score = 0)
        }

        val score = scorer.scoreLettersWord(normalized.length)
        return WordValidationResult(normalizedWord = normalized, isValid = true, failure = null, score = score)
    }
}

class ConundrumValidator {
    fun isCorrect(guess: String, answer: String): Boolean {
        return WordNormalizer.normalize(guess) == WordNormalizer.normalize(answer)
    }
}